src/Bicep.LangServer/Handlers/BicepDecompileCommandHandler.cs

```diff
@@ -0,0 +1,191 @@
+
+        public BicepDecompileCommandHandler(
+            ISerializer serializer,
+            IFeatureProviderFactory featureProviderFactory,
+            INamespaceProvider namespaceProvider,
+            IModuleRegistryProvider registryProvider,
+            ILanguageServerFacade server,
+            ITelemetryProvider telemetryProvider,
+            IApiVersionProviderFactory apiVersionProviderFactory,
+            IBicepAnalyzer bicepAnalyzer,
+            IFileResolver fileResolver)
+            : base(LangServerConstants.DecompileCommand, serializer)
+        {
+            this.telemetryHelper = new TelemetryAndErrorHandlingHelper<BicepDecompileCommandResult>(server.Window, telemetryProvider);
+            this.templateDecompiler = new TemplateDecompiler(featureProviderFactory, namespaceProvider, fileResolver, registryProvider, apiVersionProviderFactory, bicepAnalyzer);
+        }
+
+        public override Task<BicepDecompileCommandResult> Handle(BicepDecompileCommandParams parameters, CancellationToken cancellationToken)
+        {
+            return telemetryHelper.ExecuteWithTelemetryAndErrorHandling(() =>
+            {
+                if (parameters.jsonUri is null)
+                {
+                    throw new ArgumentNullException(nameof(parameters.jsonUri));
+                }
+
+                return Task.FromResult(Decompile(parameters.jsonUri.GetFileSystemPath()));
+            });
+        }
+
+        private (BicepDecompileCommandResult result, BicepTelemetryEvent? successTelemetry) Decompile(string jsonPath)
+        {
+            StringBuilder output = new StringBuilder();
+
+            Uri jsonUri = new Uri(jsonPath, UriKind.Absolute);
+
+            Uri? bicepUri;
+            ImmutableDictionary<Uri, string>? filesToSave;
+            try
+            {
+                // Decompile
+                Log(output, $"Decompiling {jsonPath} into Bicep...");
+                (bicepUri, filesToSave) = templateDecompiler.DecompileFileWithModules(jsonUri, PathHelper.ChangeToBicepExtension(jsonUri));
+            }
+            catch (Exception ex)
+            {
+                var message = $"Decompilation failed. Please fix the following problems and try again:\n{ex.Message}";
+                Log(output, message);
+                throw telemetryHelper.CreateException(
+                    message,
+                    BicepTelemetryEvent.DecompileFailure(message),
+                    new BicepDecompileCommandResult(output.ToString(), message)
+                );
+            }
+
+            // Determine output files to save
+            Trace.TraceInformation($"Decompilation main output: {bicepUri.LocalPath}");
+            Trace.TraceInformation($"Decompilation all files to save: {string.Join(", ", filesToSave.Select(kvp => kvp.Key.LocalPath))}");
+            string? outputFolder = Path.GetDirectoryName(bicepUri.LocalPath);
```

Comment:
Could you please move this before where it is used(ln 158) for readability?

Reply:
[sw/decompile2 d9b3c902b] CR fixes